sdk/tools/cabman/cabinet.h

```diff
@@ -35,19 +36,128 @@
 #define strcasecmp _stricmp
 #define strdup _strdup
 
+typedef HANDLE FILEHANDLE;
+
 #define AllocateMemory(size) HeapAlloc(GetProcessHeap(), 0, size)
 #define FreeMemory(buffer) HeapFree(GetProcessHeap(), 0, buffer)
-#define FILEHANDLE HANDLE
 #define CloseFile(handle) CloseHandle(handle)
+
+inline LONG GetSizeOfFile(FILEHANDLE handle)
+{
+    ULONG size = GetFileSize(handle, NULL);
+    if (size == INVALID_FILE_SIZE)
+        return -1;
+
+    return size;
+}
+
+inline bool ReadFileData(FILEHANDLE handle, void* buffer, ULONG size, PULONG bytesread)
+{
+    return ReadFile(handle, buffer, size, (LPDWORD)bytesread, NULL) != 0;
+}
+
+inline bool WriteFileData(FILEHANDLE handle, void* buffer, ULONG size, PULONG byteswritten)
+{
+    return WriteFile(handle, buffer, size, (LPDWORD)byteswritten, NULL) != 0;
+}
+
+inline bool IsFileHandleValid(FILEHANDLE handle)
+{
+    return handle != INVALID_HANDLE_VALUE;
+}
+
+inline FILEHANDLE OpenFileForReading(char *FileName)
+{
+    return CreateFile(FileName, GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+}
+
+inline FILEHANDLE OpenFileForWriting(char *FileName)
+{
+    return CreateFile(FileName, GENERIC_WRITE, 0, NULL, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, NULL);
+}
+
+inline bool MoveFilePointer(FILEHANDLE handle, ULONG position, ULONG *newoffset = NULL)
+{
+    ULONG pos = SetFilePointer(handle, position, NULL, FILE_BEGIN);
+
+    if (pos == INVALID_SET_FILE_POINTER)
+        return false;
+
+    if (newoffset != NULL)
+        *newoffset = pos;
+
+    return true;
+}
+
+inline bool SkipFileBytes(FILEHANDLE handle, ULONG bytestoskip)
+{
+    return SetFilePointer(handle, bytestoskip, NULL, FILE_CURRENT) != INVALID_SET_FILE_POINTER;
+}
+
 #else
+
 #define DIR_SEPARATOR_CHAR '/'
 #define DIR_SEPARATOR_STRING "/"
 
+typedef FILE* FILEHANDLE;
+
 #define AllocateMemory(size) malloc(size)
 #define FreeMemory(buffer) free(buffer)
 #define CloseFile(handle) fclose(handle)
-#define FILEHANDLE FILE*
-#endif
+
+inline LONG GetSizeOfFile(FILEHANDLE handle)
+{
+    LONG size;
+    fseek(handle, 0, SEEK_END);
+    size = ftell(handle);
+    fseek(handle, 0, SEEK_SET);
```

Comment:
I'm a bit skeptical about this... Wouldn't it be better to save current position, query for full size, and restore saved position?
So that, it would be consistent with Win32 behaviour.

Reply:
Indeed, the call to fseek() in line 113 actually does nothing and looks strange. But currently I'm trying to not mix tasks together, so I kept the implementation untouched only removing _WIN32 ifdef's.

Reply:
Ditto.